compiler/resolution/ResolutionCandidate.cpp

```diff
@@ -671,6 +671,18 @@ bool ResolutionCandidate::checkResolveFormalsWhereClauses() {
                              NULL,
                              formalIsParam) == false) {
         return false;
+      } else if (strcmp(fn->name,"init") == 0 &&
+                 formal->hasFlag(FLAG_ARG_THIS)) {
+        AggregateType* ft = toAggregateType(formal->getValType());
+        AggregateType* at = toAggregateType(actual->getValType());
+
+        // Do not allow dispatch for 'this' formal with initializers
+        //
+        // Types may only mismatch if the formal's type is an instantiation of
+        // the actual's type.
+        if (ft != at && ft->isInstantiatedFrom(at) == false) {
```

Comment:
Note that for initializers on generic types, the "this" argument might itself be generic until we have resolved Phase 1 of that initializer - so you might need another check for when the actual passed is an instantiation of the formal.  E.g.
```chapel
var foo: Foo(int, real); // Here we obtain an instantiation for Foo

// But we have no idea if the initializers for Foo have been used yet,
// so they might still be generic
foo = new Foo(...);
```

Reply:
I think the function signature of an initializer is guaranteed to be resolved by this point (at least today it is).

Reply:
Nah.  We mark the "this" argument with FLAG_DELAY_GENERIC_EXPANSION and instantiate everything else.  Then, when we go through resolution of the body, if we encounter the primitive PRIM_INIT_FIELDS, we get to [this point](https://github.com/chapel-lang/chapel/blob/master/compiler/resolution/functionResolution.cpp#L4894) of [this function](https://github.com/chapel-lang/chapel/blob/master/compiler/resolution/functionResolution.cpp#L4843), where we instantiate the type